### PR TITLE
Add Ops.reduce_{first,last} plus tests

### PR DESCRIPTION
**Pull request by danieldk**

### CODE REVIEW

**Comment by danieldk** [comment]
Let me add docs as well.

**Comment by danieldk** [comment]
> Let me add docs as well.

Done ✅

**Comment by danieldk** [comment]
@explosion-bot please test_slow_gpu

**Comment by svlandeg** [comment]
@danieldk : could you merge in `master` and rerun the slow GPU tests?

File: thinc/layers/reduce_first.py [hunk 1]
```diff
@@ -17,19 +19,13 @@ def reduce_first() -> Model[Ragged, OutT]:
...
-    starts[1:] += Xr.lengths.cumsum()[:-1]
-    X = Xr.dataXd
-    Y = cast(OutT, X[starts])
-    x_shape = Xr.dataXd.shape
-    lengths = Xr.lengths
+    Y, starts_ends = model.ops.reduce_first(cast(Floats2d, Xr.data), Xr.lengths)
 
     array_info = ArrayInfo.from_array(Y)
 
     def backprop(dY: OutT) -> Ragged:
```
Comment (svlandeg) [inline thinc/layers/reduce_first.py:26]:
```suggestion
    def backprop(dY: OutT) -> InT:
```

File: thinc/layers/reduce_first.py [hunk 2]
```diff
@@ -17,19 +19,13 @@ def reduce_first() -> Model[Ragged, OutT]:
 def forward(
     model: Model[Ragged, OutT], Xr: Ragged, is_train: bool
 ) -> Tuple[OutT, Callable[[OutT], Ragged]]:
```
Comment (svlandeg) [inline thinc/layers/reduce_first.py:21]: We probably want to use `InT` here as well for consistency? And in the `forward` of `reduce_last`.

**Review by svlandeg — COMMENTED** [review]
Definitely makes sense to have these in `ops` like the other ones.

As this affects layers that are used in some default spaCy models, let's run the spaCy slow suite from this branch if that now works? 🤞

File: thinc/layers/reduce_first.py (hunk 2 above)
Comment (danieldk) [reply thinc/layers/reduce_first.py]: Fixed.

**Comment by danieldk** [comment]
@explosion-bot please test_slow_gpu

**Comment by danieldk** [comment]
@explosion-bot please test_slow_gpu --spacy-branch master

**Comment by danieldk** [comment]
@explosion-bot please test_slow_gpu --run-on spaCy --run-on-branch master --thinc-branch refs/pull/710/head

**Comment by danieldk** [comment]
@explosion-bot please test_slow_gpu

**Review by svlandeg — APPROVED** [review]
This looks good to merge!

**Pull request merged**